sdm/src/main/java/com/sap/cds/sdm/handler/TokenHandler.java

```
@@ -140,6 +158,107 @@ public static String getDITokenUsingAuthorities(
+      httpPost.setHeader("X-zid", getTokenFields(token).get("zid").getAsString());
+
+      String encoded =
+          java.util.Base64.getEncoder()
+              .encodeToString(
+                  (sdmCredentials.getClientId() + ":" + sdmCredentials.getClientSecret())
+                      .getBytes());
+      httpPost.setHeader("Authorization", "Basic " + encoded);
+
+      List<BasicNameValuePair> basicNameValuePairs =
+          parameters.entrySet().stream()
+              .map(entry -> new BasicNameValuePair(entry.getKey(), entry.getValue()))
+              .collect(Collectors.toList());
+      httpPost.setEntity(new UrlEncodedFormEntity(basicNameValuePairs));
+
+      HttpResponse response = httpClient.execute(httpPost);
+      String responseBody = extractResponseBodyAsString(response);
+      if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
+        logger.error("Error fetching token with JWT bearer : " + responseBody);
+      }
+      Map<String, Object> accessTokenMap = new JSONObject(responseBody).toMap();
+      cachedToken = String.valueOf(accessTokenMap.get("access_token"));
+      String expiryTime = payloadObj.get("exp").getAsString();
+      CacheKey cacheKey = new CacheKey();
+      JsonObject tenantDetails = payloadObj.get("ext_attr").getAsJsonObject();
+      String subdomain = tenantDetails.get("zdn").getAsString();
+      cacheKey.setKey(payloadObj.get("email").getAsString() + "_" + subdomain);
+      cacheKey.setExpiration(expiryTime);
+      CacheConfig.getUserTokenCache().put(cacheKey, cachedToken);
+    } catch (UnsupportedEncodingException e) {
+      throw new OAuth2ServiceException("Unexpected error parsing URI: " + e.getMessage());
+    } catch (ClientProtocolException e) {
+      throw new OAuth2ServiceException(
+          "Unexpected error while fetching client protocol: " + e.getMessage());
+    } catch (IOException e) {
+      logger.error(
```

In case of IO exception do we not need to throw any exception? We are only logging it?

---

Done!